handlers/yo/yo.go

```diff
@@ -3,3 +3,253 @@ package yo
+	}
+}
+
+// StatusMessage is our HTTP handler function for status updates
+func (h *handler) StatusMessage(statusString string, channel courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.MsgStatus, error) {
+	statusForm := &statusForm{}
+	handlers.DecodeAndValidateQueryParams(statusForm, r)
+
+	// if this is a post, also try to parse the form body
+	if r.Method == http.MethodPost {
+		handlers.DecodeAndValidateForm(statusForm, r)
+	}
+
+	// validate whether our required fields are present
+	err := handlers.Validate(statusForm)
+	if err != nil {
+		return nil, err
+	}
+
+	// get our id
+	msgStatus, found := statusMappings[strings.ToLower(statusString)]
+	if !found {
+		return nil, fmt.Errorf("unknown status '%s', must be one failed, sent or delivered", statusString)
+	}
+
+	// write our status
+	status := h.Backend().NewMsgStatusForID(channel, courier.NewMsgID(statusForm.ID), msgStatus)
+	err = h.Backend().WriteMsgStatus(status)
+	if err != nil {
+		return nil, err
+	}
+
+	return []courier.MsgStatus{status}, courier.WriteStatusSuccess(w, r, status)
+}
+
+type statusForm struct {
+	ID int64 `name:"id" validate:"required"`
+}
+
+var statusMappings = map[string]courier.MsgStatusValue{
+	"failed":    courier.MsgFailed,
+	"sent":      courier.MsgSent,
+	"delivered": courier.MsgDelivered,
+}
+
+// SendMsg sends the passed in message, returning any error
+func (h *handler) SendMsg(msg courier.Msg) (courier.MsgStatus, error) {
+
+	username := msg.Channel().StringConfigForKey(courier.ConfigUsername, "")
+	if username == "" {
+		return nil, fmt.Errorf("no username set for YO channel")
+	}
+
+	password := msg.Channel().StringConfigForKey(courier.ConfigPassword, "")
+	if password == "" {
+		return nil, fmt.Errorf("no password set for YO channel")
+	}
+
+	// build our request
+	form := url.Values{
+		"origin":       []string{strings.TrimPrefix(msg.Channel().Address(), "+")},
+		"sms_content":  []string{courier.GetTextAndAttachments(msg)},
+		"destinations": []string{strings.TrimPrefix(msg.URN().Path(), "+")},
+		"ybsacctno":    []string{username},
+		"password":     []string{password},
+	}
+
+	var status courier.MsgStatus
+	encodedForm := form.Encode()
+	sendURLs := []string{sendURL1, sendURL2, sendURL3}
+
+	for _, sendURL := range sendURLs {
+		failed := false
+		sendURL := fmt.Sprintf("%s?%s", sendURL, encodedForm)
+
+		req, err := http.NewRequest(http.MethodGet, sendURL, nil)
+		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+		rr, err := utils.MakeHTTPRequest(req)
+
+		if err != nil {
+			failed = true
+		}
+		// record our status and log
+		status = h.Backend().NewMsgStatusForID(msg.Channel(), msg.ID(), courier.MsgErrored)
+		status.AddLog(courier.NewChannelLogFromRR(msg.Channel(), msg.ID(), rr))
+		if err != nil {
+			return status, err
+		}
+
+		responseQS, err := url.ParseQuery(string(rr.Body))
+
+		if err != nil {
+			failed = true
+		}
+
+		if failed == false && rr.StatusCode != 200 && rr.StatusCode != 201 {
```

Comment:
Maybe see if you can refactor this a bit with `continue`s instead of checking failed everywhere